pandas/core/internals/blocks.py

```diff
@@ -1007,6 +1008,9 @@ def coerce_to_target_dtype(self, other):
         # if we cannot then coerce to object
         dtype, _ = infer_dtype_from(other, pandas_dtype=True)
 
+        if isinstance(self.dtype, StringDtype):
```

Comment:
specifically handling things is not great here. i would rather add an is_extension_dtype check below

Reply:
I try to use is_extension_array_dtype, but Categorical and DatetimeArray don't work. So I add `not is_categorical_dtype` and `not is_datetime` in the same conditional expression.